nautobot/extras/models/metadata.py

```diff
@@ -135,3 +138,108 @@ def save(self, *args, **kwargs):
     def delete(self, *args, **kwargs):
         # TODO: should we behave like CustomFieldChoice and raise a ProtectedError if this choice is in use anywhere?
         super().delete(*args, **kwargs)
+
+
+@extras_features(
+    "custom_validators",
+    "graphql",
+    "webhooks",
+)
+class ObjectMetadata(ChangeLoggedModel, BaseModel):
+    metadata_type = models.ForeignKey(
+        to=MetadataType,
+        on_delete=models.PROTECT,
+        related_name="object_metadatas",
+    )
+    contact = models.ForeignKey(
+        to=Contact,
+        on_delete=models.PROTECT,
+        related_name="object_metadatas",
+        blank=True,
+        null=True,
+    )
+    team = models.ForeignKey(
+        to=Team,
+        on_delete=models.PROTECT,
+        related_name="object_metadatas",
+        blank=True,
+        null=True,
+    )
+    scoped_fields = JSONArrayField(
+        base_field=models.CharField(
+            max_length=CHARFIELD_MAX_LENGTH,
+        ),
+        help_text="List of scoped fields, only direct fields on the model",
+    )
+    value = models.JSONField(
+        blank=True,
+        null=True,
+        help_text="Relevant data value to an object field or a set of object fields",
+    )
+    assigned_object_type = models.ForeignKey(to=ContentType, on_delete=models.SET_NULL, null=True, related_name="+")
+    assigned_object_id = models.UUIDField(db_index=True)
+    assigned_object = GenericForeignKey(ct_field="assigned_object_type", fk_field="assigned_object_id")
+
+    natural_key_field_names = ["pk"]
+    documentation_static_path = "docs/user-guide/platform-functionality/metadata.html"
+
+    class Meta:
+        ordering = ["metadata_type"]
+
+        indexes = [
+            models.Index(
+                name="assigned_object",
+                fields=["assigned_object_type", "assigned_object_id"],
+            ),
+            models.Index(
+                name="assigned_object_contact",
+                fields=["assigned_object_type", "assigned_object_id", "contact"],
+            ),
+            models.Index(
+                name="assigned_object_team",
+                fields=["assigned_object_type", "assigned_object_id", "team"],
+            ),
+        ]
+
+    def __str__(self):
+        if self.contact:
+            return f"{self.contact}: {self.metadata_type} - {self.assigned_object} - {self.value}"
+        else:
+            return f"{self.team}: {self.metadata_type} - {self.assigned_object} - {self.value}"
```

Comment:
Contact and team are both specific types of metadata, so probably this should be something like:

```suggestion
        if self.contact:
            return f"{self.metadata_type} - {self.assigned_object} - {self.contact}"
        elif self.team:
            return f"{self.metadata_type} - {self.assigned_object} - {self.team}"
        else:
            return f"{self.metadata_type} - {self.assigned_object} - {self.value}"
```

I wonder if we should make the `value` database field into a private field (e.g. `_value`) and have `value` instead be an `@property` that returns either the contact, the team, or the `_value` as appropriate?

Reply:
Implemented `value` helper function